File: practica1/src/gestorAplicacion/prestamo/Prestamo.java
package gestorAplicacion.prestamo;

import java.io.Serializable;
import java.time.*;
import java.util.ArrayList;

import gestorAplicacion.obras.Libro;
import gestorAplicacion.obras.Publicacion;
import gestorAplicacion.obras.Revista;
import gestorAplicacion.personas.EstudianteProfesor;
import gestorAplicacion.personas.Externo;
import gestorAplicacion.personas.Persona;

public class Prestamo implements Serializable {
	
	//ATRIBUTOS DE CLASE
	private static final long serialVersionUID = 1L;
	private static ArrayList<Prestamo> lista = new ArrayList<>();
	private final static LocalDate fechaActual= LocalDate.now();	
	
	
	//ATRIBUTOS DE INSTANCIA
	private int id;
	private LocalDate inicio;
	private LocalDateTime fin;
	private Publicacion publicacion;
	private Persona usuario;
	
	//CONSTRUCTORES
	public Prestamo(EstudianteProfesor usuario, Publicacion publicacion,int id,LocalDate inicio) { // constructor para los usuarios de la universidad
		this.usuario=usuario;
		this.publicacion=publicacion;
		this.id=id;
		this.inicio=inicio;
		usuario.getPrestamos().add(this);
		lista.add(this);
	}
	public Prestamo(Externo usuario, Publicacion publicacion,int id,LocalDate inicio) { // constructor para los usuarios externos de la universidad
		this.usuario=usuario;
		this.publicacion=publicacion;
		this.id=id;
		this.inicio=inicio;
		usuario.getPrestamos().add(this);
		lista.add(this);
	}
	
	//METODOS
	 public void determinarFin(EstudianteProfesor usuario, Libro publicacion) {
		if ( publicacion.getTipo() == Libro.tipoLibro.RESERVA) {
			this.fin= this.inicio.plusDays(1).atTime(LocalTime.of(12, 0));
		}
		else if( publicacion.getTipo() == Libro.tipoLibro.COLECCION_GENERAL) {
			this.fin= this.inicio.plusDays(30).atTime(LocalTime.of(12, 0));
		}else {}
	}
	 
	 public void determinarFin(EstudianteProfesor usuario, Revista publicacion) {
		this.fin= this.inicio.plusDays(15).atTime(LocalTime.of(12, 0));
		}
	 
	 public void determinarFin(Externo usuario, Libro publicacion) {
		 if ( publicacion.getTipo() == Libro.tipoLibro.COLECCION_GENERAL) {
			 this.fin= this.inicio.plusDays(15).atTime(LocalTime.of(12, 0));
		 }
	}
	 
	 public void determinarFin(Externo usuario, Revista publicacion) {
		 this.fin= this.inicio.plusDays(7).atTime(LocalTime.of(12, 0));
	}
	 public String mostrarInfo() {
		 return " DETALLE DEL PRESTAMO " + "\n" + "Publicacion: " + this.publicacion.getNombre() + "\n" + "Usuario: " + this.usuario.getNombre() + "\n" 
					+ "ID: " + this.id + "\n" + "Fecha de inicio: " + this.inicio + "\n" + "Fecha de vencimiento: " + this.fin;
	 }
	
	
	//GETTERS SETTERS
	public static ArrayList<Prestamo> getLista() {
		return lista;
	}
	public static void setLista(ArrayList<Prestamo> lista) {
		Prestamo.lista = lista;
	}
	public static LocalDate getFechaactual() {
		return fechaActual;
	}
	
	public int getId() {
		return this.id;
	}
	
	public Publicacion getPublicacion() {
		return this.publicacion;
	}
	
	public LocalDateTime getFin() {
		return this.fin;
	}
	public void setFin(LocalDateTime fin) {
		this.fin = fin;
		
	}
	
	

}

File: practica2-python/gestorAplicacion/prestamo/Prestamo.py
import datetime
from gestorAplicacion.obras.Libro import Libro, tipoLibro
from gestorAplicacion.obras.Revista import Revista
from gestorAplicacion.obras.Folleto import Folleto


class Prestamo:

    #Atributo clase
    
    _lista=[]
    _fechaActual= datetime.datetime.now()

    #constructor
    
    def __init__(self, id=0,inicio='...', publicacion=None, usuario=None):
        Prestamo._lista.append(self)
        self.id = id
        self.inicio = inicio
        self.fin = None
        self.publicacion = publicacion
        self.usuario = usuario
        self.activo = False
        
        

    #metodos

    def determinarFinInterno(self): 
        if isinstance(self.publicacion,Libro):
            if self.publicacion.getTipo() == tipoLibro.r:
                self.fin = self.inicio + datetime.timedelta(days=1)
                self.fin = self.fin.replace(hour=12)
            elif self.publicacion.getTipo() == tipoLibro.cg:
                self.fin = self.inicio + datetime.timedelta(days=30)
                self.fin = self.fin.replace(hour=12)
            else:
                return "Los libros de Tesis y/o Investigacion no estan disponibles para prestamo"

        elif isinstance( self.publicacion ,Revista):
            self.fin = self.inicio + datetime.timedelta(days=15)
            self.fin = self.fin.replace(hour=12)
        
        elif isinstance(self.publicacion,Folleto):
            return "Los Folletos no estan disponibles para prestamo"

    
    def determinarFinExterno(self): 
        if isinstance(self.publicacion,Libro):
            if self.publicacion.getTipo() == tipoLibro.r:
                return "Los usuarios externos de la universidad no pueden prestar libros de reserva"
                
            elif self.publicacion.getTipo() == tipoLibro.cg:
                self.fin = self.inicio + datetime.timedelta(days=15)
                self.fin = self.fin.replace(hour=12)
            else:
                return "Los libros de Tesis y/o Investigacion no estan disponibles para prestamo"

        elif isinstance( self.publicacion ,Revista):
            self.fin = self.inicio + datetime.timedelta(days=7)
            self.fin = self.fin.replace(hour=12)
        
        elif isinstance(self.publicacion,Folleto):
            return "Los Folletos no estan disponibles para prestamo"

    def mostrarInfo(self):
        return f'DETALLE DEL PRESTAMO: \n Material prestado: {self.publicacion.getNombre()} CP: {self.publicacion.getCodigo()} \n Usuario: {self.usuario.getNombre()} CC: {self.usuario.getId()} \n Fecha inicio: {self.inicio} \n Fecha vencimiento: {self.fin}'

    @classmethod
    def mostrarRegistros(cls):
        # from gestorAplicacion.obras.Publicacion import Publicacion
        # from gestorAplicacion.personas.Persona import Persona
        c = "PRESTAMOS REALIZADOS: \n"
        # for x in Prestamo._lista:
        #     if isinstance(x,Prestamo):
        #         publicacion = x.getPublicacion()
        #         usuario = x.getUsuario()
        #         if isinstance(publicacion,Publicacion) and isinstance(usuario,Persona):
        #             c = c + f'Prestamo con ID: {x.getId()}  Publicacion: {publicacion.getNombre()} Usuario: {usuario.getNombre()} Inicio: {x.getInicio()} Fin: {x.getFin()}\n'
        return c



    @classmethod
    def getLista(cls):
        return Prestamo._lista
    @classmethod
    def setLista(cls,lista):
        Prestamo._lista = lista

    @classmethod
    def getFechaActual(cls):
        return Prestamo._fechaActual

    def getId(self):
        return self.id
    def setId(self,i):
        self.id = i

    def getInicio(self):
        return self.inicio
    def setInicio(self,i):
        self.inicio =i

    def getFin(self):
        return self.fin
    def setFin(self,f):
        self.fin = f

    def getUsuario(self):
        return self.usuario
    def setUsuario(self,p):
        self.usuario = p

    def getPublicacion(self):
        return self.publicacion
    def setPublicacion(self,p):
        self.publicacion = p

    def isActivo(self):
        return self.activo
    def setActivo(self,v):
        self.activo = v